feature_tests/js/api/Bar.mjs

```diff
@@ -27,8 +27,11 @@ export class Bar {
         
         this.#ptr = ptr;
         this.#selfEdge = selfEdge;
-        // Unconditionally register to destroy when this object is ready to garbage collect.
-        Bar_box_destroy_registry.register(this, this.#ptr);
+        
+        // Are we being borrowed? If not, we can register.
+        if (this.#selfEdge.length === 0) {
+            Bar_box_destroy_registry.register(this, this.#ptr);
+        }
```

Comment:
Thought: This condition, `this.#selfEdge.length === 0`, is a strange case that seems brittle. I spent about half an hour looking carefuly at this with @ambiguousname and I think I understand where it comes from: the array is usually empty, except when the returned struct is owned by someone else, as it is in `Bar::foo`. It raises a question of whether this is the right place to add the struct to the finalization registry; would it make more sense to do it at the call site where the struct is being constructed? Any time we have weird conditions like this, we get further away from my vision of having reference-counted JS-WASM memory management (instead of the garbage-collected edge arrays).

Reply:
This is what we do in Dart as well. We should autogenerate a comment saying what is going on here.

I think this is the right place to register the struct, half the time the callsite doesn't know if things are being borrowed or not either (since it can be in another struct, where that depends on the caller).

> my vision of having reference-counted JS-WASM memory management (instead of the garbage-collected edge arrays).

I don't understand this vision